### ISSUE TITLE
Add rule for warning against non-final static fields

### ISSUE THREAD

**Issue by scottsteen** [spotbugs/spotbugs#2418]
Mutating non-final static fields could lead to thread-safety issues. The proposed rule would issue a warning for any fields which are `static` but not `final`.

Example

```java
class ViolatingClass {
  // ok fields
  private static final String finalStatic_pass = "";
  private final String finalInstance_pass = ""; 
  private String nonFinalInstance_pass;

  // violating fields
  private static String nonFinalStatic_fail;
}
```

Happy to submit a PR if you'd like this rule.

**Comment by scottsteen** [comment]
I've just noticed that `MutableStaticFields` exists, but ignores `private` & `package-private` fields & those not in `public` classes. Is this intentional? We recently had an issue where a `private static` field was mutated in a constructor of a `private static` nested class (the developer's intention was for it to be `private final`).

Could this rule cover fields of _all_ access modifiers in _all_ classes?

**Comment by gtoison** [comment]
There are many cases where people do want to mutate static fields, such a rule would probably generate many false positive.
Spotbugs is not smart enough to know if the intent of the developer was to mutate the field or to make it final.

**Comment by scottsteen** [comment]
Thanks for the consideration. I get that there may be some cases where mutable static state is necessary (some lazy instantiation implementations, for example). 

Would it be considered at all as a new detector with low priority reporting in the "Dodgy Code" category? More as an indicator that the developer needs to be aware that they may need to do extra work to make sure the implementation is thread-safe. Akin to the `EQ_UNUSUAL` & `NS_NON_SHORT_CIRCUIT` "questionable code" rules. A separate Bug Pattern would allow folks to opt out completely via the exlcusions mechanism.

If not, thanks for the time & the tool!

**Comment by gtoison** [comment]
I think that rule might be too opinionated for SpotBugs, the project tends to focus on "bugs", not "code smells".

**Comment by hazendaz** [comment]
Agreed on focus on bugs.  Personally, I disable IDE setup on ever suggesting where to use final or static.  It's incredibly noisy otherwise.  This rule would be too.

Some tooling like open rewrite is getting better at such concerns though.  And honestly, it's better these days that a bot fix the issues for users and explain why as it's more likely applied.  I find most tools even spotbugs get ignored by far too many devs already.  Making it light up with very subjective items even with good intent will just reinforce ignoring the tooling.

Best bet for you would be to take eclipse for example. Set those settings in source cleanup then run source cleanup and look at the diff.  That may be eye opening...

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by gtoison** [comment]
I'll close this issue as it seems to be outside of the project scope, please don't hesitate to comment here or to open an other issue if you feel otherwise!
